builder/config.go

```diff
@@ -0,0 +1,358 @@
+	fields := []struct {
+		re       string
+		dest     *string
+		required bool
+	}{
+		// [Builder]
+		{`^BUNDLE_DIR\s*=\s*`, &config.Builder.BundleDir, true}, //Note: Can be removed once UseNewChrootBuilder is obsolete
+		{`^CERT\s*=\s*`, &config.Builder.Cert, true},
+		{`^SERVER_STATE_DIR\s*=\s*`, &config.Builder.ServerStateDir, true},
+		{`^VERSIONS_PATH\s*=\s*`, &config.Builder.VersionPath, true},
+		{`^YUM_CONF\s*=\s*`, &config.Builder.DNFConf, true},
+		// [Swupd]
+		{`^BUNDLE\s*=\s*`, &config.Swupd.Bundle, false},
+		{`^CONTENTURL\s*=\s*`, &config.Swupd.ContentURL, false},
+		{`^FORMAT\s*=\s*`, &config.Swupd.Format, true},
+		{`^VERSIONURL\s*=\s*`, &config.Swupd.VersionURL, false},
+		// [Server]
+		{`^debuginfo_banned\s*=\s*`, &config.Server.DebugInfoBanned, false},
+		{`^debuginfo_lib\s*=\s*`, &config.Server.DebugInfoLib, false},
+		{`^debuginfo_src\s*=\s*`, &config.Server.DebugInfoSrc, false},
+		// [Mixer]
+		{`^LOCAL_BUNDLE_DIR\s*=\s*`, &config.Mixer.LocalBundleDir, false},
+		{`^LOCAL_REPO_DIR\s*=\s*`, &config.Mixer.LocalRepoDir, false},
+		{`^LOCAL_RPM_DIR\s*=\s*`, &config.Mixer.LocalRPMDir, false},
+	}
+
+	for _, h := range fields {
+		r := regexp.MustCompile(h.re)
+		for _, i := range lines {
+			if m := r.FindIndex([]byte(i)); m != nil {
+				*h.dest = i[m[1]:]
+			}
+		}
+	}
+
+	if config.Mixer.LocalBundleDir == "" {
+		pwd, err := os.Getwd()
+		if err != nil {
+			return err
+		}
+		config.Mixer.LocalBundleDir = filepath.Join(pwd, "local-bundles")
+		fmt.Printf("WARNING: LOCAL_BUNDLE_DIR not found in builder.conf. Falling back to %q.\n", config.Mixer.LocalBundleDir)
+		fmt.Println("Please set this value to the location you want local bundle definition files to be stored.")
+	}
+
+	return nil
+}
+
+func (config *MixConfig) expandEnv() error {
+	re := regexp.MustCompile(`\$\{?([[:word:]]+)\}?`)
+	rv := reflect.ValueOf(config).Elem()
+
+	for i := 0; i < rv.NumField(); i++ {
+		sectionV := rv.Field(i)
+
+		for j := 0; j < sectionV.NumField(); j++ {
+			val := sectionV.Field(j).String()
+			matches := re.FindAllStringSubmatch(val, -1)
+
+			for _, s := range matches {
+				if _, ok := os.LookupEnv(s[1]); !ok {
+					return errors.Errorf("buildconf contains an undefined environment variable: %s\n", s[1])
+				}
+			}
+
+			sectionV.Field(j).SetString(os.ExpandEnv(val))
+		}
+
+	}
+
+	return nil
+}
+
+func (config *MixConfig) validate() error {
+	rv := reflect.ValueOf(config).Elem()
+
+	for i := 0; i < rv.NumField(); i++ {
+		sectionT := reflect.TypeOf(rv.Field(i).Interface())
+		sectionV := rv.Field(i)
+
+		for j := 0; j < sectionT.NumField(); j++ {
+			tag, ok := sectionT.Field(j).Tag.Lookup("required")
+
+			if ok && tag == "true" && sectionV.Field(j).String() == "" {
+				return errors.Errorf("Missing required field in config file: %s", sectionT.Field(j).Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Convert parses an old config file and converts it to TOML format
+func (config *MixConfig) Convert(filename string) error {
```

Comment:
My only nitpick so far...calling convert again borks the builder.conf if it has already been converted/is in any format other than the original format. Granted, I don't expect people to run it twice, but you never know. It might be nice to see if we can already read it in the new format and just return nil in that case.

Reply:
That is one of the reasons I was talking about adding a version number, so we can identify what config we are parsing. Is this case it won't really be a problem since calling convert on a new config will fail on the first parse step.

Reply:
Agreed to both. Good simple next step :)